core/src/main/java/org/elasticsearch/index/similarity/SimilarityService.java

```diff
@@ -122,15 +131,15 @@ public SimilarityProvider getSimilarity(String name) {
     }
 
     Similarity getDefaultSimilarity() {
-        return defaultSimilarity;
+        return defaultSimilarity.get();
     }
 
     static class PerFieldSimilarity extends PerFieldSimilarityWrapper {
 
-        private final Similarity defaultSimilarity;
+        private final SimilarityProvider defaultSimilarity;
         private final MapperService mapperService;
 
-        PerFieldSimilarity(Similarity defaultSimilarity, Similarity baseSimilarity, MapperService mapperService) {
+        PerFieldSimilarity(SimilarityProvider defaultSimilarity, Similarity baseSimilarity, MapperService mapperService) {
```

Comment:
this change worries me. We now have a set of mutable objects (SimilarityProvider) and we do either tons of volatile reads or access synchronized blocks. We do this bascially for each createWeight call and even further 2  createWeight calls can get 2 different similarities. I think we shouldn't implement the update the way we do it here but rather keep all the provider immutable and if there is an update to any of the similarity settings we rebuild the entire set of providers. this would guarantee that we have a consistent view on all the values as well as a consistent view across all users ie one search sees one set of similarities? does this make sense?

it also has the advantage that we only have a single update method for these values per index